gen/field.go

```diff
@@ -179,9 +182,25 @@ func compileJSONTag(f *compile.FieldSpec, name string, opts ...string) *structta
 	// should be omitted from the encoding if the field has an empty value,
 	// defined as false, 0, a nil pointer, a nil interface value, and any empty
 	// array, slice, map, or string.
-	if (isReferenceType(f.Type) || isStructType(f.Type) || isPrimitiveType(f.Type)) &&
-		!f.Required && !t.HasOption("omitempty") {
-		t.Options = append(t.Options, "omitempty")
+	//
+	// If the field is marked with "!omitempty", then "omitempty" will not be added.
+	// If both "!omitempty" and "omitempty" are present, then "omitempty" is removed.
+	hasNotOmitempty := t.HasOption(notOmitempty)
+	hasOmitempty := t.HasOption(omitempty)
+	if (isReferenceType(f.Type) || isStructType(f.Type) || isPrimitiveType(f.Type)) && !f.Required {
```

Comment:
nit: can move these into the if block
```suggestion
	if (isReferenceType(f.Type) || isStructType(f.Type) || isPrimitiveType(f.Type)) && !f.Required {
		hasNotOmitempty := t.HasOption(notOmitempty)
		hasOmitempty := t.HasOption(omitempty)
```